Replace AbstractSecretKeyLoader with ISecretKeyProvider, Fixes AB#3297746

p3dr0rv opened the pull request:
In the next PR:

* We are replacing AbstractSecretKeyLoader with ISecretKeyProvider.
* We are also delegating the secret key generation to a new class to separate concerns and improve code readability.

All these changes lay the groundwork for the next PR, where the new implementation will be introduced and the feature flag will be used to switch between the two implementations.

[AB#3297746](https://identitydivision.visualstudio.com/fac9d424-53d2-45c0-91b5-ef6ba7a6bf26/_workitems/edit/3297746)

p3dr0rv committed:
add interface ISecretKeyLoader

p3dr0rv committed:
nits

p3dr0rv committed:
nits

p3dr0rv committed:
nits

p3dr0rv committed:
fix interface error

p3dr0rv committed:
fix error

p3dr0rv committed:
fix tests

p3dr0rv committed:
Merge branch 'dev' into pedroro/interface-for-secretkeyloader

p3dr0rv committed:
Update AndroidWrappedKeyLoader.java

rpdome commented on common/src/androidTest/java/com/microsoft/identity/common/crypto/AndroidWrappedKeyLoaderTest.java:
```diff
@@ -113,7 +113,7 @@ public void testGenerateKey() throws ClientException {
         final SecretKey secretKey = keyLoader.generateRandomKey();
 
-        Assert.assertEquals(AES256KeyLoader.AES_ALGORITHM, secretKey.getAlgorithm());
+        Assert.assertEquals(keyLoader.getSecretKeyGenerator().getKeyAlgorithm(), secretKey.getAlgorithm());
```
the goal here is to compare with the actual value (constant), not from what we get from the keyloader object.

Perhaps revert all the changes in this class?

rpdome commented on common/src/main/java/com/microsoft/identity/common/crypto/AndroidWrappedKeyLoaderFactory.kt:
```diff
@@ -0,0 +1,76 @@
+ * specific loader implementation should be used, allowing for runtime switching between
+ * different implementations without affecting client code.
+ */
+object AndroidWrappedKeyLoaderFactory {
```
Rather than a keyloader factory, it looks like you need a factory for storage layer instead?

two ways to tackle this

1. the new storage layer can decrypt (read) with both old and new keyloader. it would only write with the new keyloader

2.  the storage re-encrypts its data with the new keyloader, and use the new keyloader to read/write.

rpdome replied:
If we simply use the flight to switch here - it would lead to user's data loss (since they can't decrypt existing data)

rpdome replied:
e.g. put the flight in AndroidBrokerEncryptionManagerWithKeyStoreWrappedKey's getKeyLoaderForEncryption() and getKeyLoaderForDecryption(), if you're going with the first route.

p3dr0rv replied:
Yeah, I’m planning to address this in the next PR, where I’ll be using strategy 1, see -> 
https://github.com/AzureAD/microsoft-authentication-library-common-for-android/blob/db8781af8f7746c3df4a51ee78f0fde2d82a661a/common/src/main/java/com/microsoft/identity/common/crypto/AndroidKeyStoreRsaKekManager.kt#L111

Instead of handling it at the storage layer as you mentioned, I’m just plugging in a new keyloader. This new keyloader will be easier to update if we need to introduce a new algorithm, whether for the key itself or the key-encryption-key.

p3dr0rv committed:
Refactor key generation to use a shared AES256SecretKeyGenerator instance

p3dr0rv committed:
Merge branch 'dev' into pedroro/interface-for-secretkeyloader

p3dr0rv committed:
Merge branch 'dev' into pedroro/interface-for-secretkeyloader

p3dr0rv committed:
Merge branch 'dev' into pedroro/interface-for-secretkeyloader

mohitc1 commented on common/src/main/java/com/microsoft/identity/common/crypto/AndroidWrappedKeyLoaderFactory.kt:
```diff
@@ -0,0 +1,76 @@
+    fun createWrappedKeyLoader(
+        keyIdentifier: String,
+        fileName: String,
+        context: android.content.Context
```
>android.content.Context [](http://example.com/codeflow?start=17&length=23)

nit: Context

mohitc1 commented on common4j/src/main/com/microsoft/identity/common/java/crypto/key/ISecretKeyGenerator.kt:
```diff
@@ -0,0 +1,64 @@
+     * This should be compatible with cryptographic providers, such as
+     * those used with KeyGenerator.getInstance(algorithm).
+     */
+    val keyAlgorithm: String
```
do we need these?

p3dr0rv replied:
these are just informational getters, removed

mohitc1 commented on common/src/main/java/com/microsoft/identity/common/crypto/AndroidWrappedKeyLoader.java:
```diff
@@ -177,12 +185,11 @@ public synchronized SecretKey getKey() throws ClientException {
 
-    @Override
     @NonNull
     protected SecretKey generateRandomKey() throws ClientException {
```
>protected [](http://example.com/codeflow?start=4&length=9)

private?

mohitc1 commented on common/src/main/java/com/microsoft/identity/common/crypto/AndroidWrappedKeyLoaderFactory.kt:
```diff
@@ -0,0 +1,76 @@
+ * different implementations without affecting client code.
+ */
+object AndroidWrappedKeyLoaderFactory {
+    const val WRAPPED_KEY_KEY_IDENTIFIER: String = "A001"
```
>WRAPPED_KEY_KEY_IDENTIFIER [](http://example.com/codeflow?start=14&length=26)

do you need this here? (maybe for next pr)

p3dr0rv replied:
move for next pr

mohitc1 commented on common4j/src/main/com/microsoft/identity/common/java/flighting/CommonFlight.java:
```diff
@@ -124,7 +124,12 @@ public enum CommonFlight implements IFlightConfig {
+    /**
+     * Flight to enable the new Android wrapped key loader.
+     */
+    ENABLE_NEW_ANDROID_WRAPPED_KEY_LOADER("EnableNewAndroidWrappedKeyLoader", true);
```
Do you need this in this PR?

You could add this in next PR and factory class

p3dr0rv replied:
move for next pr

mohitc1 commented on common4j/src/main/com/microsoft/identity/common/java/crypto/key/ISecretKeyLoader.kt:
```diff
@@ -0,0 +1,88 @@
+ *
+ * [ISecretKeyLoader] provides a consistent abstraction layer for cryptographic key operations
+ */
+interface ISecretKeyLoader {
```
>ISecretKeyLoader [](http://example.com/codeflow?start=10&length=16)

Given it's functionality, I think, a better name would be ISecretKeyProvider.

your call.

p3dr0rv replied:
change  name

mohitc1 reviewed (approved):
Left some suggestions

p3dr0rv committed:
Merge branch 'dev' into pedroro/interface-for-secretkeyloader

p3dr0rv committed:
Refactor key loading mechanism to use ISecretKeyProvider interface, replacing ISecretKeyLoader. Update related classes and tests to ensure compatibility with new key provider structure. Remove deprecated AndroidWrappedKeyLoaderFactory and adjust encryption manager implementations accordingly.

p3dr0rv committed:
Remove keySize and keyAlgorithm properties from ISecretKeyGenerator and AES256SecretKeyGenerator. Update tests accordingly to reflect these changes.

p3dr0rv committed:
fix issues

p3dr0rv committed:
add changelog

rpdome commented on common/src/main/java/com/microsoft/identity/common/crypto/AndroidAuthSdkStorageEncryptionManager.java:
```diff
@@ -54,26 +53,27 @@ public class AndroidAuthSdkStorageEncryptionManager extends StorageEncryptionMan
 
-    private final PredefinedKeyLoader mPredefinedKeyLoader;
-    private final AndroidWrappedKeyLoader mKeyStoreKeyLoader;
+    private final PredefinedKeyProvider mPredefinedKeyLoader;
```
nit: rename variables accordingly?

rpdome commented on common/src/main/java/com/microsoft/identity/common/crypto/AndroidWrappedKeyLoader.java:
```diff
@@ -71,7 +71,15 @@
  * before it get saved to the file.
  */
-public class AndroidWrappedKeyLoader extends AES256KeyLoader {
+public class AndroidWrappedKeyLoader extends AbstractAES256SecretKeyProvider {
```
nit: rename class, loader -> provider?

rpdome commented on common4j/src/main/com/microsoft/identity/common/java/crypto/key/AbstractAES256SecretKeyProvider.java:
```diff
@@ -22,42 +22,21 @@
+ * for concrete implementations that handle the loading of AES-256 secret keys from
+ * various sources.
+ */
+public abstract class AbstractAES256SecretKeyProvider implements ISecretKeyProvider {
```
doesn't look like this abstract class is overriding anything. do we need this?

if all we need are constants/static object - we can get rid of this class and
- expose KEY_ALGORITHM in AES256SecretKeyGenerator
- add a static getInstance() in AES256SecretKeyGenerator() to replace AES_256_KEY_GENERATOR

p3dr0rv replied:
done

rpdome reviewed (approved):
Left a few suggestions. otherwise lgtm.

p3dr0rv committed:
Refactor key management to implement ISecretKeyProvider interface, replacing AbstractAES256SecretKeyProvider. Update related classes and tests for compatibility. Remove deprecated classes.

p3dr0rv committed:
fix tests

p3dr0rv committed:
Refactor key loader to key provider in encryption manager

- Updated the AndroidAuthSdkStorageEncryptionManagerTest to replace instances of ISecretKeyLoader with ISecretKeyProvider.
- Refactored StorageEncryptionManager to use key providers instead of key loaders for encryption and decryption processes.
- Introduced MockAES256KeyProvider and MockAES256KeyProviderWithGetKeyError classes to facilitate testing with the new key provider interface.
- Adjusted StorageEncryptionManagerTest to utilize the new MockAES256KeyProvider for testing encryption and decryption functionalities.
- Ensured all references to key loaders in the codebase are replaced with key providers for consistency and clarity.

p3dr0rv committed:
Rename method to setIgnoreKeyProviderNotFoundError for consistency in AuthenticationSettings

p3dr0rv committed:
Rename key loader variable to key provider for consistency in AndroidAuthSdkStorageEncryptionManager and update related test assertion.

p3dr0rv committed:
Add TODO comment for wrapCipher usage and simplify getKey method in PredefinedKeyProvider
